layers/ipfix.go

```diff
@@ -0,0 +1,773 @@
+	// IANA enterprise number [IANA-PEN] of the authority defining the
+	// Information Element identifier in this Template Record.
+	EnterpriseNumber uint32
+}
+
+// Len returns the length in octet for Field Specifier
+//   * 4 for IANA Information Element
+//   * 8 for enterprise-specific Information Element
+func (fs IPFIXFieldSpecifier) Len() int {
+	if fs.EnterpriseBit {
+		return 2 + 2 + 4
+	}
+	return 2 + 2
+}
+
+func (fs *IPFIXFieldSpecifier) decodeFromBytes(data *[]byte) error {
+	if len(*data) < fs.Len() {
+		*data = []byte{}
+		return fmt.Errorf("invalid field specifier length")
+	}
+
+	*data, fs.InformationElementID = (*data)[2:], binary.BigEndian.Uint16((*data)[:2])
+	*data, fs.FieldLength = (*data)[2:], binary.BigEndian.Uint16((*data)[:2])
+
+	// If the Enterprise bit is one, the Information Element identifier
+	// identifies an enterprise-specific Information Element, and the Enterprise
+	// Number field MUST be present.
+	if fs.InformationElementID&EnterpriseBit > 0 {
+		fs.EnterpriseBit = true
+		fs.InformationElementID ^= EnterpriseBit
+		// check length again as it depends if it is an entreprise field specifier
+		if len(*data) < fs.Len()-4 {
+			*data = []byte{}
+			return fmt.Errorf("invalid enterprise number specifier length")
+		}
+
+		*data, fs.EnterpriseNumber = (*data)[4:], binary.BigEndian.Uint32((*data)[:4])
+	}
+	return nil
+}
+
+// IPFIXDataSet is a set of Data Records (RFC 7011 section 3.3 and 3.4.3)
+//
+//     0                   1                   2                   3
+//     0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |   Set ID = Template ID        |          Length               |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |   Record 1 - Field Value 1    |   Record 1 - Field Value 2    |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |   Record 1 - Field Value 3    |             ...               |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |   Record 2 - Field Value 1    |   Record 2 - Field Value 2    |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |   Record 2 - Field Value 3    |             ...               |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |   Record 3 - Field Value 1    |   Record 3 - Field Value 2    |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |   Record 3 - Field Value 3    |             ...               |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+//    |              ...              |      Padding (optional)       |
+//    +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+type IPFIXDataSet struct {
+	// Set header
+	Header IPFIXSetHeader
+	// Data records in the set
+	Records []IPFIXDataRecord
+	// Raw bytes
+	Bytes []byte
```

Comment:
what is the purpose of maintaining raw bytes? Is this for the case, where we do not know the template? If so, can we save the space by just store raw bytes for that scenario and not for cases, where we know the template?

Reply:
yes I though about that optimization but did not implemented it yet. Let me add to the todo for later